### PR TITLE
Disable infinispan realm and user cache for map storage tests

### LINKED ISSUE #11213: Disable infinispan realm and user cache for map storage tests
The new storage will not use the old caching layer. Currently, we still have it enabled in the testsuite. Disabling it may uncover some issues in the storage. We should disable it so that we can fix these hidden issues.

### PR DESCRIPTION
**Pull request by vramik**
Closes #11213

### CODE REVIEW

**Comment by hmlnarik** [comment]
The hot rod tests take too long to finish.
@mhajas @martin-kanis Could you investigate [near caching feature](https://docs.jboss.org/infinispan/12.0/apidocs/org/infinispan/client/hotrod/configuration/RemoteCacheConfigurationBuilder.html#nearCacheMode(org.infinispan.client.hotrod.configuration.NearCacheMode)) in Infinispan?

File: model/map/src/main/java/org/keycloak/models/map/realm/MapRealmAdapter.java [hunk 1]
```diff
@@ -848,7 +862,7 @@ public void updateAuthenticatorConfig(AuthenticatorConfigModel model) {
         entity.getAuthenticatorConfig(model.getId())
                         .ifPresent(oldAC -> {
                             oldAC.setAlias(model.getAlias());
-                            oldAC.setConfig(model.getConfig());
+                            oldAC.setConfig(model.getConfig() == null ? null : new HashMap<>(model.getConfig()));
```
Comment (mhajas) [inline model/map/src/main/java/org/keycloak/models/map/realm/MapRealmAdapter.java:865]: This shouldn't be necessary. Each implementation of `setConfig` should make a deep copy of the parameter. 

Current implementation looks like this:
```java
@SuppressWarnings("unchecked") @Override public void setConfig(java.util.Map<String,String> p0) {
        p0 = p0 == null ? null : new java.util.HashMap<>(p0);
        if (p0 != null) {
            p0.values().removeIf(org.keycloak.models.map.common.UndefinedValuesUtils::isUndefined);
            if (org.keycloak.models.map.common.UndefinedValuesUtils.isUndefined(p0)) p0 = null;
        }
        updated |= ! Objects.equals(fConfig, p0);
        fConfig = p0;
    }
```

File: model/map/src/main/java/org/keycloak/models/map/realm/MapRealmAdapter.java [hunk 2]
```diff
@@ -899,7 +919,7 @@ public void updateRequiredActionProvider(RequiredActionProviderModel model) {
                             oldRAP.setPriority(model.getPriority());
                             oldRAP.setEnabled(model.isEnabled());
                             oldRAP.setDefaultAction(model.isDefaultAction());
-                            oldRAP.setConfig(model.getConfig());
+                            oldRAP.setConfig(model.getConfig() == null ? null : new HashMap<>(model.getConfig()));
```
Comment (mhajas) [inline model/map/src/main/java/org/keycloak/models/map/realm/MapRealmAdapter.java:922]: This should not be necesasary.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/SAMLServletAdapterTest.java [hunk 3]
```diff
@@ -939,7 +939,7 @@ public void salesPostSigTest() {
     @Test
     // https://issues.jboss.org/browse/KEYCLOAK-3971
     public void salesPostSigTestUnicodeCharacters() {
-        final String username = "ěščřžýáíRoàåéèíñòøöùüßÅÄÖÜ";
+        final String username = "ěščřžýáíroàåéèíñòøöùüßåäöü";
```
Comment (mhajas) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/SAMLServletAdapterTest.java:942]: Why is this necessary?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/realm/RealmTest.java [hunk 4]
```diff
@@ -685,6 +686,7 @@ public static void assertRealm(RealmRepresentation realm, RealmRepresentation st
 
     @Test
     public void clearRealmCache() {
+        Assume.assumeTrue("Realm cache disabled.", isRealmCacheEnabled());
```
Comment (mhajas) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/realm/RealmTest.java:689]: Just out of curiosity, it is not necessary to fix this. Is there some simple summary of why the tests need this?

**Review by mhajas — CHANGES_REQUESTED** [review]
Thank you @vramik, the changes look good to me. However, could you have a look at copying collections when setting into an entity? I thought we don't need it as each parameter is copied automatically if it is immutable.

File: model/jpa/src/main/java/org/keycloak/models/jpa/RealmAdapter.java [hunk 5]
```diff
@@ -1797,6 +1797,9 @@ public Stream<AuthenticatorConfigModel> getAuthenticatorConfigsStream() {
 
     @Override
     public RequiredActionProviderModel addRequiredActionProvider(RequiredActionProviderModel model) {
+        if (getRequiredActionProviderByAlias(model.getAlias()) != null) {
```
[the remark this answers is not in the document]
Comment (vramik) [reply model/jpa/src/main/java/org/keycloak/models/jpa/RealmAdapter.java:1800]: Yes. 

Without it we allow to store multiple required action providers with the same alias. Then when we call `realmModel.getRequiredActionProviderByAlias(String)` we obtain *first* occurrence (see `jpa.RealmAdapter` implementation) of the object that match the alias. It results in non-deterministic behavior when sometimes you got one object, another time different one. With JPA this bug is kind of hidden as you usually get the same order of objects stored in the database so you get the same object most of the time. 

But with map it became visible when some tests failed intermittently when cache is disabled. 

It doesn't make sense to have registered the required action more than once anyway.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/SAMLServletAdapterTest.java (hunk 3 above)
Comment (vramik) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/SAMLServletAdapterTest.java:942]: `MapUserProvider` doesn't use `username.toLowerCase()` when querying users. Therefore the test doesn't work when there are some upper case characters in username. 

This test is about testing username with special characters and username lettercase will be handled within https://github.com/keycloak/keycloak/issues/10245.

Another possibility would be to skip the test for now till https://github.com/keycloak/keycloak/issues/10245 is done, but I prefer this solution. wdyt?

File: model/jpa/src/main/java/org/keycloak/models/jpa/RealmAdapter.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (mhajas) [reply model/jpa/src/main/java/org/keycloak/models/jpa/RealmAdapter.java:1800]: Ahh I see, I expected there is some unique key for this in JPA. Thanks for the explanation.

File: model/map/src/main/java/org/keycloak/models/map/realm/MapRealmAdapter.java [hunk 6]
```diff
@@ -521,6 +524,14 @@ public void addRequiredCredential(String cred) {
...
             throw new RuntimeException("Unknown credential type " + cred);
         }
+        if (entity.getRequiredCredentials() != null) {
+            Optional<MapRequiredCredentialEntity> existing = entity.getRequiredCredentials().stream()
+                    .filter(credential -> Objects.equals(model.getType(), credential.getType()))
+                    .findFirst();
+            if (existing.isPresent()) {
+                throw new ModelDuplicateException("A Required Credential with given type already exists.");
+            }
+        }
```
Comment (vramik) [inline model/map/src/main/java/org/keycloak/models/map/realm/MapRealmAdapter.java:534]: @mhajas  This piece of code could be replaced by
```
if (getRequiredCredentialsStream().anyMatch(credential -> Objects.equals(model.getType(), credential.getType()))) {
    throw new ModelDuplicateException("A Required Credential with given type already exists.");
}
```

which is much more readable but `getRequiredCredentialsStream()` transforms all entities to models and I wanted to avoid that. 

Another option would be to add new method `Optional<MapRequiredCredentialEntity> getRequiredCredentialByType(String type);` to `MapRealmEntity`. 

WDYT?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/SAMLServletAdapterTest.java (hunk 3 above)
Comment (mhajas) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/SAMLServletAdapterTest.java:942]: Is #10245 somehow relevant to querying users, isn't it only about stored lettercase? If we choose to store username preserving the lettercase, do we want to allow two users with usernames `admin` and `Admin`? I am asking because currently, I see that we query usernames with `ILIKE` twice and once with `EQ`, should we switch `EQ` to `ILIKE`? Would it solve the issue with this test?

File: model/map/src/main/java/org/keycloak/models/map/realm/MapRealmAdapter.java (hunk 6 above)
Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/realm/MapRealmAdapter.java:534]: I like method `getRequiredCredentialByType(String type)`

Comment (hmlnarik) [reply model/map/src/main/java/org/keycloak/models/map/realm/MapRealmAdapter.java:534]: I like it as well, only in a separate PR to think it through.
For example of the questions - this pattern may be common to other places, should we unify? Should this method even be available in an `Entity`?

This PR should introduce only minimal changes, and that the suggestion above satisfies.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/SAMLServletAdapterTest.java (hunk 3 above)
Comment (hmlnarik) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/SAMLServletAdapterTest.java:942]: > I see that we query usernames with `ILIKE` twice and once with `EQ`, should we switch `EQ` to `ILIKE`?

There definitely needs to be a consistency in the search by username. If that is the case, could you please introduce an issue with direct references to the code to fix this? 

The other question is how each storage treats this `EQ` / `ILIKE` inside it. Some are case-sensitive and some case-insensitive. This is treated in #10245

File: model/map/src/main/java/org/keycloak/models/map/realm/MapRealmAdapter.java (hunk 6 above)
Comment (vramik) [reply model/map/src/main/java/org/keycloak/models/map/realm/MapRealmAdapter.java:534]: so leaving it as is for now and I'll create new enhancement issue to address it, thanks

Comment (hmlnarik) [reply model/map/src/main/java/org/keycloak/models/map/realm/MapRealmAdapter.java:534]: Sorry, I was not clear. Please use the form you suggested above.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/SAMLServletAdapterTest.java (hunk 3 above)
Comment (mhajas) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/SAMLServletAdapterTest.java:942]: I created an issue here: https://github.com/keycloak/keycloak/issues/11602, feel free to add more details

File: model/map/src/main/java/org/keycloak/models/map/realm/MapRealmAdapter.java (hunk 6 above)
Comment (vramik) [reply model/map/src/main/java/org/keycloak/models/map/realm/MapRealmAdapter.java]: Created https://github.com/keycloak/keycloak/issues/11603

**Review by mhajas — APPROVED** [review]
Thanks for the changes @vramik

**Review by hmlnarik — APPROVED** [review]
Approving per @mhajas 's approval
